rollup/l1/msg_storage.go

```diff
@@ -0,0 +1,262 @@
+package l1
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/core/types"
+	"github.com/scroll-tech/go-ethereum/log"
+)
+
+const (
+	defaultFetchInterval = 5 * time.Second
+)
+
+type MsgStorageState struct {
+	StartBlockHeader *types.Header
+	EndBlockHeader   *types.Header
+}
+
+type MsgStorage struct {
+	state MsgStorageState
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	msgs                  *common.ShrinkingMap[uint64, storedL1Message]
+	reader                *Reader
+	unsubscribeTracker    func()
+	newChainNotifications []newChainNotification
+
+	msgsMu   sync.RWMutex
+	notifsMu sync.Mutex
+}
+
+func NewMsgStorage(ctx context.Context, tracker *Tracker, reader *Reader) (*MsgStorage, error) {
+	if tracker == nil || reader == nil {
+		return nil, fmt.Errorf("failed to create MsgStorage, reader or tracker is nil")
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	msgStorage := &MsgStorage{
+		ctx:    ctx,
+		cancel: cancel,
+		msgs:   common.NewShrinkingMap[uint64, storedL1Message](1000),
+		reader: reader,
+	}
+	msgStorage.unsubscribeTracker = tracker.Subscribe(LatestChainHead, func(old, new []*types.Header) {
+		msgStorage.notifsMu.Lock()
+		defer msgStorage.notifsMu.Unlock()
+		msgStorage.newChainNotifications = append(msgStorage.newChainNotifications, newChainNotification{old, new})
+	})
+
+	msgStorage.Start()
+	return msgStorage, nil
+}
+
+func (ms *MsgStorage) Start() {
+	log.Info("starting MsgStorage")
+	go func() {
+		fetchTicker := time.NewTicker(defaultFetchInterval)
+		defer fetchTicker.Stop()
+
+		for {
+			select {
+			case <-ms.ctx.Done():
```

Comment:
Why need to specifically handle `<-ms.ctx.Done()` and the subsequent select handle this again